### ISSUE TITLE
Warning: The plugin local_auth_android requires Android SDK version 34

### ISSUE THREAD

**Issue by edisonlsm** [flutter/flutter#143605]
As recommended when using endorsed packages, our app depends only on `local_auth`, which depends on local_auth_android by itself, with a ^1.0.0 clause.

`local_auth_android` updated the compileSdk from 33 to 34 on a **patch version** (1.0.36 to 1.0.37), so when we ran a fresh build on our CI  the new version was automatically installed by flutter pub and the project started demanding a new compileSdk version even though I did not changed any dependency on **pubspec.yaml**.

Shouldn't an update this relevant be made only on a major (or at least a minor) version? Specially on an endorsed packaged, which repos should not worry too much about. We were about to release a new version to the app stores when we were surprised by this warning. 

I know bumping the compileSdk version of the app is not a bad thing, but this is a change on the code that demands some testings on the app... It is something we would like to do when planned and not rushed out because of a patch update of an endorsed package.

Should we expect more updates like this from the official packages? Should I lock the versions of **all** the endorsed packages to make sure my codebase stay stable?

---

What package does this bug report belong to?

local_auth

What target platforms are you seeing this bug on?

Android

Have you already upgraded your packages?

Yes

<details><summary>pubspec.lock</summary>

```lock
local_auth:
    dependency: transitive
    description:
      name: local_auth
      sha256: "27679ed8e0d7daab2357db6bb7076359e083a56b295c0c59723845301da6aed9"
      url: "https://pub.dev"
    source: hosted
    version: "2.1.8"
  local_auth_android:
    dependency: transitive
    description:
      name: local_auth_android
      sha256: "3bcd732dda7c75fcb7ddaef12e131230f53dcc8c00790d0d6efb3aa0fbbeda57"
      url: "https://pub.dev"
    source: hosted
    version: "1.0.37"
  local_auth_ios:
    dependency: transitive
    description:
      name: local_auth_ios
      sha256: "6dde47dc852bc0c8343cb58e66a46efb16b62eddf389ce103d4dacb0c6c40c71"
      url: "https://pub.dev"
    source: hosted
    version: "1.1.7"
```
</details>

**Comment by stuartmorgan** [comment]
> the project started demanding a new compileSdk version

Can you elaborate on what you mean by "demanded"? You've referred to this several times as a warning; warnings are advisory, which is what distinguishes them from errors. Did you have an actual build failure? If so, what exactly was the failure log?

> Shouldn't an update this relevant be made only on a major (or at least a minor) version?

We generally follow semver, and a `compileSdk` is a small change that does not affect the API surface.

> Should we expect more updates like this from the official packages?

Yes, this is our standard practice for updating `compileSdk`, and has been for some time; we have not received any feedback on in before that I am aware of. If this is problematic for your use case, please provide more details on specifically what issues this caused, and why you believe that it should be considered a breaking change.

> Should I lock the versions of **all** the endorsed packages to make sure my codebase stay stable?

`pubspec.lock` already locks the versions of all of your dependencies, both direct and transitive. The Dart [guidance on `pubspec.lock`](https://dart.dev/guides/libraries/private-files#pubspec-lock) has been consistent for quite some time that application developers should be checking in a lock file. If you are unexpectedly getting new versions of dependencies in your CI, that would indicate that your application is not currently set up to follow existing best practice.

**Comment by edisonlsm** [comment]
Yes, a warning is usually just a warning, but the language on this warning (which I know is not from you guys but from Android SDK itself) got me confused. "Requires" does not sound so optional for me, that's why we stopped our build, when reading this warning here we could not be sure that everything would still work as expected.
![image](https://github.com/flutter/flutter/assets/4001673/7900088f-00e8-472e-82e7-97cb1a9f8ddd)


I know a `compileSdk` flag change does not change the APIs from the package itself, but since it changes the entire build environment around it doesn't it mean it is a kind of a big change? I mean, the methods stay the same, the exposed API from the package stay the same but there is a chance that some behaviour changed on the OS itself, no? I always see this kind of update as a big one. I always bump minor/major when I update dependencies, Flutter itself or even some configs on my code because I see them as potential breaking changes. Maybe I am being too literal with "breaking"?


But I will take the blame on the last one, we should have checked pubspec.lock in a long time ago and it would have avoided this issue on our code. I am still surprised with this upgrade behaviour, but now that I know how we can avoid it we will adopt it here and check pubspec.lock in and lock everything. No more surprises!

**Comment by stuartmorgan** [comment]
> Yes, a warning is usually just a warning, but the language on this warning (which I know is not from you guys but from Android SDK itself) got me confused.

The messages in your screenshot are coming from the `flutter` tool, so we could potentially reword them.

Routing to the Android team to consider. I may be misremembering, but IIRC the current wording is because if a plugin actually uses features from a newer compile SDK, the build will actually fail; if that's the case we could consider make this phrasing more conditional (e.g., "One or more plugins expects a higher Android SDK version. If building plugins fails, you may be able to resolve the problem by adding [...]").

**Comment by gmackall** [comment]
> > Yes, a warning is usually just a warning, but the language on this warning (which I know is not from you guys but from Android SDK itself) got me confused.
> 
> The messages in your screenshot are coming from the `flutter` tool, so we could potentially reword them.

Yes, they come [from here](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L860C29-L860C58). I think if we could figure out how to get the warning level on the logger we use (I don't think it is currently enabled) we could make this yellow instead, which would probably be an improvement.

**Comment by lukehutch** [comment]
> We generally follow semver, and a `compileSdk` is a small change that does not affect the API surface.

The problem with the error message is that the end user has no idea whether the Flutter developers are actually using features of the new API or not. It could be dangerous to ignore these warnings (not to mention that showing big red warnings that can be safely ignored leads to attention fatigue, which causes users to eventually ignore warnings that are in fact serious).

If it were never dangerous to ignore the warning, then the warning would not be shown. Therefore, I would not call this a small change. Also, since increasing the compileSdk exposes a whole new set of API changes, I would not say that this does not affect the API surface -- it does affect the *transitive* API surface.

I really think the semver major version should be increased whenever `compileSdk` is increased, so that projects that depend upon Flutter packages using `^` can continue to run `pub upgrade` without suddenly being faced with these ugly error messages, or potential breakage.

As a second alternative, I suggest that `compileSdk` be specified in `pubspec.yaml`. This could then be used as an additional constraint by `pub` when resolving dependencies, so that you can only depend upon libraries with the same or lower API level.

**Comment by lukehutch** [comment]
> Routing to the Android team to consider. I may be misremembering, but IIRC the current wording is because if a plugin actually uses features from a newer compile SDK, the build will actually fail; if that's the case we could consider make this phrasing more conditional (e.g., "One or more plugins expects a higher Android SDK version. If building plugins fails, you may be able to resolve the problem by adding [...]").

You are neglecting behavioral changes. The ABI of a library can stay the same while its API contract changes, because an API level increase introduces behavioral changes.

**Comment by stuartmorgan** [comment]
> Also, since increasing the compileSdk exposes a whole new set of API changes, I would not say that this does not affect the API surface -- it does affect the _transitive_ API surface.

There is no "transitive API surface"; the API surface of our plugins is the public Dart API it exposes.

> I really think the semver major version should be increased whenever `compileSdk` is increased

This has been discussed in the past, and it is not a policy we will be implementing.

> You are neglecting behavioral changes. The ABI of a library can stay the same while its API contract changes, because an API level increase can introduce behavioral changes.

Any change to a package can introduce behavioral changes (many do, intentionally), but that's not how semver is defined. Anyone who wants to treat any change that could potentially change behavior of the package as breaking can exactly pin all packages, which has the same effect.

**Comment by lukehutch** [comment]
> I may be misremembering, but IIRC the current wording is because if a plugin actually uses features from a newer compile SDK, the build will actually fail; if that's the case we could consider make this phrasing more conditional

How about simply not even showing the warning, if you are convinced if a library uses newer APIs, the build will break?

The warning looks very serious, and the only way to avoid it is by pinning the dep to a specific earlier version, which is a pain, because you have to go through the changelog or bisect the version number until you find the version right before the API bump. But if this warning really won't affect anyone until new API is actually used, then it really should not be shown.

Maybe if the build of a dependency fails, then this warning can be tacked on the end as a suggestion as to why the build may have broken.

**Comment by stuartmorgan** [comment]
> How about simply not even showing the warning, if you are convinced if a library uses newer APIs, the build will break?

Because a build break message caused by a too-low compile SDK will not have a clear, actionable message, and we have seen that they are frequently very difficult for many developers to debug/understand.

> The warning looks very serious

Thus the existence of this issue, which specifically treats that as a bug to be fixed.

> the only way to avoid it is by pinning the dep to a specific earlier version

That is definitely to the only way; the more common way for someone to avoid it is to follow the steps that the warning message provides, and update to a newer compile SDK.

> Maybe if the build of a dependency fails, then this warning can be tacked on the end as a suggestion as to why the build may have broken.

We only do that when we can clearly associate the actual failure with a specific remedy. There are many reasons a build can fail, and we can't easily identify (e.g., via regex of output) when this message would be the right one to show. We know from experience that showing the *wrong* suggestion at the end of build error output causes significant confusion.

**Comment by camsim99** [comment]
I think the action item here is to rework the Flutter tool warning as mentioned above by @stuartmorgan and @gmackall, by exploring the possibility of changing the color to indicate it is a warning and/or re-word the message to indicate it may/may not be necessary. 

> Because a build break message caused by a too-low compile SDK will not have a clear, actionable message, and we have seen that they are frequently very difficult for many developers to debug/understand.

To corroborate Stuart's point, [this issue](https://github.com/flutter/flutter/issues/91771) is what prompted adding the warning and is the type of issue this warning will help avoid.

> > Maybe if the build of a dependency fails, then this warning can be tacked on the end as a suggestion as to why the build may have broken.
> 
> We only do that when we can clearly associate the actual failure with a specific remedy. There are many reasons a build can fail, and we can't easily identify (e.g., via regex of output) when this message would be the right one to show. We know from experience that showing the wrong suggestion at the end of build error output causes significant confusion.

It's unfortunate that there's no real way to detect when plugins use APIs from new SDK versions, but since we can't, I agree that the warning is important to keep for that case, especially since changing it _should_ be harmless.

On that note, I acknowledge that the other concern here is that unintended behavioral changes may come with a compileSdk version bump, but Android has set the expectation that your app's behavior _should_ remain the same when the compileSdk version is bumped. If there were ignored but announced changes that came with a new SDK version, then not catching this before making the version bump would be more of a reflection of a plugin's test coverage versus an issue in our versioning system, if anything.

**Comment by lukehutch** [comment]
> That is definitely to the only way; the more common way for someone to avoid it is to follow the steps that the warning message provides, and update to a newer compile SDK.

Doesn't that cut out many potential new users of your app, because they won't even be able to download the app from the Google Play store if their API level is not high enough? API level 34 is currently only on 12.6% of phones! API level 33 is on 41.4%.

https://apilevels.com/

In maybe 5 years, all the major phone manufacturers will have realized that they have to keep up to date with the yearly API bump on Sept 1st, or they will lose users. But until then, increasing the minimum API level for your app will have serious app visibility consequences.

**Comment by stuartmorgan** [comment]
> Doesn't that cut out many potential new users of your app, because they won't even be able to download the app from the Google Play store if their API level is not high enough?

No; what you are describing is `minSdkVersion`, which is unrelated to this warning. I recommend reading [this](https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd).

**Comment by lukehutch** [comment]
> No; what you are describing is `minSdkVersion`, which is unrelated to this warning. I recommend reading [this](https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd).

I don't understand. How does an app not break at runtime if `compileSdk` > `minSdk`, the app is running on a phone with only `minSdk` installed, and the app makes use of new API methods that aren't in `minSdk`? Ignoring the subtlety of behavioral changes of `targetSdk`, it seems like `minSdk` is the only version that actually matters, since it determines what will actually break or not break on devices.
